### PR TITLE
cd

### PR DESCRIPTION
**Pull request by harou24**

### CODE REVIEW

File: src/builtin_cd.c [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		if (cmd->argv->argv[1][0] == '.')
+			return (go_to_path(cmd->argv->argv[1]));	/*check_for_real_path*/
+		else if (ft_strcmp(cmd->argv->argv[1], "-"))
+		{
+			int ret = go_to_prev_dir();
+			ft_dprintf(cmd->fds[FD_OUT], "%s\n", env_get_current_dir());
+			return (ret);
+		}
+		else
+			return (go_to_path(cmd->argv->argv[1]));
```
Comment (s-t-a-n) [inline src/builtin_cd.c]: put in function 'go_to_path'

don't fiddle arround with "." or "..", they are real files (== folders)

check if you can go to the folder, then set the oldpwd   and pwd (pwd -> realpath) and change into the directory

File: src/builtin_cd.c [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	if (cmd->argv->argc == 1)
+		return(go_to_home());
+	else if (cmd->argv->argc == 2)
+	{
+		if (cmd->argv->argv[1][0] == '.')
+			return (go_to_path(cmd->argv->argv[1]));	/*check_for_real_path*/
+		else if (ft_strcmp(cmd->argv->argv[1], "-"))
+		{
+			int ret = go_to_prev_dir();
+			ft_dprintf(cmd->fds[FD_OUT], "%s\n", env_get_current_dir());
```
Comment (s-t-a-n) [inline src/builtin_cd.c]: cd never prints the directory

**Review by s-t-a-n — CHANGES_REQUESTED** [review]

File: src/builtin_cd.c (hunk 2 above)
Comment (harou24) [reply src/builtin_cd.c:33]: In the man it says that :

 "-         When a <hyphen-minus> is used as the operand, this shall be equivalent to the command:
 97 
 98                      cd "$OLDPWD" && pwd
 99 
100                  which changes to the previous working directory and then writes its name."

Comment (harou24) [reply src/builtin_cd.c:33]: In the man it says
-         When a <hyphen-minus> is used as the operand, this shall be equivalent to the command:
 97 
 98                      cd "$OLDPWD" && pwd
 99 
100                  which changes to the previous working directory and then writes its name.

Comment (harou24) [reply src/builtin_cd.c]: In the man it says:
" -         When a <hyphen-minus> is used as the operand, this shall be equivalent to the command:
 97 
 98                      cd "$OLDPWD" && pwd
 99 
100                  which changes to the previous working directory and then writes its name.
"

Comment (s-t-a-n) [reply src/builtin_cd.c]: i stand corrected!

File: src/builtin_cd.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+static int	__go_to_home()
+{
+	return (fs_change_dir(env_get_home()));
+}
+
+static t_bool	__update_dir_change(t_command *cmd)
+{
+	return (env_set_s("OLDPWD", env_get_s("PWD")) && 
+			env_set_s("PWD", cmd->argv->argv[1]));
```
Comment (s-t-a-n) [inline src/builtin_cd.c]: not much use of getters in env_access if you set them manually 

also is there a reason why you are passing t_cmd and not char * ?

File: src/builtin_cd.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	{
+		dir_change = __go_to_prev_dir();
+		ft_dprintf(cmd->fds[FD_OUT], "%s\n", env_get_current_dir());
+	}
+	else
+		dir_change = fs_change_dir(cmd->argv->argv[1]);
+	if (dir_change == DIR_HAS_CHANGED)
+		__update_dir_change(cmd);
+	return (dir_change);
+}
```
Comment (s-t-a-n) [inline src/builtin_cd.c]: again, what has __go_to_path has to do with t_command ??

File: src/builtin_cd.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	if (cmd->argv->argc == 1)
+		dir_change = __go_to_home();
+	else if (cmd->argv->argc == 2)
+		dir_change = __go_to_path(cmd);
+	else
+	{
+		/*handle errors*/
+		return (-1);
+	}
+	if (dir_change != DIR_HAS_CHANGED)
```
Comment (s-t-a-n) [inline src/builtin_cd.c]: i get what you are trying to do here but this makes it way harder to read. 

change 'dir_change' to error

also, if you were to use that type of boolean, use 'dir_has_changed'

File: src/builtin_cd.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		return (-1);
+	exit(EXIT_SUCCESS);
```
Comment (s-t-a-n) [inline src/builtin_cd.c]: why return -1 and not exit with the error code ?

File: src/builtin_cd.c [hunk 7]
```diff
@@ -0,0 +1,58 @@
+#include "bash_ops.h"
+#include "env_access.h"
+#include "filesystem_traversal.h"
+#include "ft_printf.h"
+#include "path.h"
+#include <stdlib.h>
```
Comment (s-t-a-n) [inline src/builtin_cd.c:7]: order your includes, this is a mess

system libs on top
then user defined standard libs
then the libs specific to this file

**Review by s-t-a-n — CHANGES_REQUESTED** [review]

**Review by s-t-a-n — APPROVED** [review]
awesome! very clean

do mind that we have to split up the functions into seperate files, so maybe make most of the functions non-static (you can leave them in the file)

**Pull request merged**